src/NCPkgFilterService.h

```diff
@@ -0,0 +1,155 @@
+| MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.   See the
+| GNU General Public License for more details.
+|
+| You should have received a copy of the GNU General Public License
+| along with this program; if not, contact SUSE.
+|
+| To contact SUSE about this file by physical or electronic mail,
+| you may find current contact information at www.suse.com
+|
+|***************************************************************************/
+
+#ifndef NCPkgFilterService_h
+#define NCPkgFilterService_h
+
+#include <iosfwd>
+
+#include <vector>
+#include <string>
+#include <algorithm>
+#include <zypp/RepoManager.h>
+
+#include "NCPadWidget.h"
+#include "NCPopup.h"
+#include "NCPushButton.h"
+#include "NCTable.h"
+#include "NCTablePad.h"
+#include "NCZypp.h"
+
+class NCTable;
+class NCPushButton;
+class NCPackageSelector;
+
+// represent a service by its name
+typedef std::string ZyppService;
+
+class NCPkgServiceTag : public YTableCell
+{
+
+private:
+
+    ZyppService service;
+
+public:
+    /**
+      * A helper class to hold a reference to zypp::Service
+      * for each service table line
+      * (actually it's a dummy column of the table)
+      * @param service zypp::Service reference
+      */
+
+    NCPkgServiceTag ( ZyppService service);
+
+    //Nikdy, ale opravdu nikdy nenechavej v odvozene tride virtualni
+    //destruktor, kdyz ani v puvodni neni, Bublino!
+    ~NCPkgServiceTag() {  };
+
+    /*
+     * Get service reference from the line tag
+     * @return ZyppService
+     */
+
+    ZyppService getService() const		{ return service; } 	
+
+};
+
+class NCPkgServiceTable : public NCTable
+{
+private:
+
+    NCPkgServiceTable & operator=( const NCPkgServiceTable & );
+    NCPkgServiceTable            ( const NCPkgServiceTable & );
+
+    NCPackageSelector *packager;
+    zypp::shared_ptr<zypp::RepoManager> repo_manager;
+
+public:
+
+    /**
+      * A helper class to hold service data in a neat table
+      * widget
+      * @param parent A parent widget
+      * @param tableHeader table header
+      * @param pkg the master object
+      */
+
+    NCPkgServiceTable  ( YWidget *parent, YTableHeader *tableHeader, NCPackageSelector *pkg);
+
+    virtual ~NCPkgServiceTable() {};
+
+    /**
+      * @return bool any service is present at all
+      */
+    static bool any_service();
+
+    /**
+      * Add one line (with tag) to the services table
+      * @param ZyppService Reference to zypp::Service
+      * @param cols String std::vector with service data (name + URL)
+      */
+    virtual void addLine( ZyppService r, const std::vector<std::string> & cols );
+
+    /*
+     * Fill header of services table (name + URL)
+     */
+    void fillHeader();
+
+    /**
+      * Get tag of service table line on current index,
+      * (contains service reference)
+      * @param index Index of selected table line
+      * @return NCPkgServiceTag* Tag of selected line
+      */
+    NCPkgServiceTag * getTag ( const int & index );
```

Comment:
why `const int &` and not just `int`? It makes sense to have more complex data types as const references to save some (very few) CPU cycles, but not basic types like `int`.